src/hip_stream.cpp

```diff
@@ -254,14 +254,137 @@ hipError_t hipStreamGetPriority(hipStream_t stream, int* priority) {
 
 
 //---
+hipError_t setCallbackPacket(hsa_queue_t* queue,
+                             uint64_t& index, uint64_t& nextIndex,
+                             hsa_barrier_and_packet_t** barrier1,
+                             hsa_barrier_and_packet_t** barrier2){
+
+    if(queue == nullptr ||
+       barrier1 == nullptr || barrier2 == nullptr)
+            return hipErrorInvalidValue;
+
+    uint64_t tempIndex = 0;
+    uint32_t mask = queue->size - 1;
+    hsa_barrier_and_packet_t* tempBarrier;
+
+    // Check for empty packets
+    do{
+        tempIndex = hsa_queue_load_write_index_scacquire(queue);
+        tempBarrier = &(((hsa_barrier_and_packet_t*)(queue->base_address))[tempIndex & mask]);
+    }while(!(tempBarrier->header & HSA_PACKET_TYPE_INVALID));
+
+    // Reserve two packets for two barriers
+    index = hsa_queue_add_write_index_scacquire(queue, 2);
+
+    if(index > mask)
+    {
+        index = 0;
+        nextIndex = 1;
+    }
+    else if(index == mask)
+        nextIndex = 0;
+    else
+        nextIndex = index + 1;
+
+    tempBarrier = new hsa_barrier_and_packet_t;
+    memset(tempBarrier, 0, sizeof(hsa_barrier_and_packet_t));
+    tempBarrier->header = HSA_PACKET_TYPE_INVALID;
+
+    // Barrier 1
+    *barrier1 = &(((hsa_barrier_and_packet_t*)(queue->base_address))[index & mask]);
+    memcpy(*barrier1,tempBarrier,sizeof(hsa_barrier_and_packet_t));
+
+    // Barrier 2
+    *barrier2 = &(((hsa_barrier_and_packet_t*)(queue->base_address))[nextIndex & mask]);
+    memcpy(*barrier2,tempBarrier,sizeof(hsa_barrier_and_packet_t));
+
+    delete tempBarrier;
+
+    return hipSuccess;
+}
+
 hipError_t hipStreamAddCallback(hipStream_t stream, hipStreamCallback_t callback, void* userData,
                                 unsigned int flags) {
+
     HIP_INIT_API(hipStreamAddCallback, stream, callback, userData, flags);
     hipError_t e = hipSuccess;
 
-    // Create a thread in detached mode to handle callback
+    if(stream == hipStreamNull)
+    {
+        ihipCtx_t* device = ihipGetTlsDefaultCtx();
+        stream = device->_defaultStream;
+    }
+
+    // Lock the stream
+    LockedAccessor_StreamCrit_t crit(stream->criticalData(), true /*unlock at destruction*/);
```

Comment:
the default value of autoUnlock argument of LockedAccessor constructor is true, so if your intention is to use the default value then it's not necessary to pass true here when you lock the stream.

Reply:
I have purposefully kept 'true' here as passing it will not have any extra cost but add better readability.